File: sdks/python/src/crashlens/_transport.py
"""Background, non-blocking HTTP transport.

A single daemon thread drains a bounded queue and POSTs one event per request.
Design rules from the brief:

* the app thread NEVER blocks: :meth:`Transport.submit` drops the newest event
  when the queue is full (warning once) rather than waiting;
* the body is gzip-compressed when it exceeds :data:`GZIP_THRESHOLD`;
* on a connection error or 5xx we retry ONCE after a short delay, then drop;
* on 429 we sleep up to :data:`MAX_RETRY_AFTER` seconds honouring ``Retry-After``
  and drop (never re-queue);
* 4xx (other than 429) drops immediately;
* nothing here ever raises into the host application.
"""

from __future__ import annotations

import gzip
import json
import logging
import queue
import threading
import time
import urllib.error
import urllib.request
import warnings
from typing import Any, Dict, Optional

logger = logging.getLogger("crashlens")

GZIP_THRESHOLD = 4096  # 4 KiB
RETRY_DELAY = 0.5  # seconds, single retry on connection error / 5xx
MAX_RETRY_AFTER = 5  # seconds, cap on honouring a 429 Retry-After

_SHUTDOWN = object()


class Transport:
    def __init__(
        self,
        url: str,
        key: str,
        *,
        timeout: float = 2.0,
        max_queue: int = 100,
    ) -> None:
        self._url = url
        self._key = key
        self._timeout = timeout
        self._queue: "queue.Queue[Any]" = queue.Queue(maxsize=max_queue)
        self._cond = threading.Condition()
        self._pending = 0
        self._closed = False
        self._warned_full = False
        self._thread = threading.Thread(
            target=self._run, name="crashlens-transport", daemon=True
        )
        self._thread.start()

    # -- producer side -----------------------------------------------------

    def submit(self, envelope: Dict[str, Any]) -> None:
        """Enqueue an event without blocking. Drops (newest) when full."""
        if self._closed:
            return
        with self._cond:
            try:
                self._queue.put_nowait(envelope)
            except queue.Full:
                if not self._warned_full:
                    warnings.warn(
                        "crashlens: event queue is full; dropping events",
                        stacklevel=2,
                    )
                    self._warned_full = True
                return
            self._pending += 1

    def flush(self, timeout: float = 5.0) -> bool:
        """Block until the queue drains or ``timeout`` elapses. Returns success."""
        deadline = time.monotonic() + timeout
        with self._cond:
            while self._pending > 0:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    return False
                self._cond.wait(remaining)
        return True

    def close(self, timeout: float = 5.0) -> None:
        """Flush, stop the worker, and join it. Idempotent."""
        if self._closed:
            return
        self._closed = True
        self.flush(timeout)
        try:
            self._queue.put_nowait(_SHUTDOWN)
        except queue.Full:
            # Make room by draining one item, then signal shutdown.
            try:
                self._queue.get_nowait()
            except queue.Empty:
                pass
            try:
                self._queue.put_nowait(_SHUTDOWN)
            except queue.Full:
                pass
        self._thread.join(timeout=timeout)

    # -- consumer side -----------------------------------------------------

    def _run(self) -> None:
        while True:
            item = self._queue.get()
            if item is _SHUTDOWN:
                return
            try:
                self._send(item)
            except Exception:  # noqa: BLE001 - transport must never crash
                logger.debug("crashlens: event send failed", exc_info=True)
            finally:
                with self._cond:
                    self._pending -= 1
                    if self._pending <= 0:
                        self._cond.notify_all()

    def _serialize(self, envelope: Dict[str, Any]) -> "tuple[bytes, dict[str, str]]":
        body = json.dumps(
            envelope, default=str, ensure_ascii=False, separators=(",", ":")
        ).encode("utf-8")
        headers = {
            "Content-Type": "application/json; charset=utf-8",
            "X-Crashlens-Key": self._key,
            "User-Agent": "crashlens-python",
        }
        if len(body) > GZIP_THRESHOLD:
            body = gzip.compress(body)
            headers["Content-Encoding"] = "gzip"
        return body, headers

    def _send(self, envelope: Dict[str, Any]) -> None:
        body, headers = self._serialize(envelope)
        for attempt in (0, 1):
            try:
                status, retry_after = self._post(body, headers)
            except (urllib.error.URLError, TimeoutError, OSError):
                # Connection-level failure: retry once, then drop.
                if attempt == 0:
                    time.sleep(RETRY_DELAY)
                    continue
                logger.debug("crashlens: connection error, dropping event")
                return
            if status == 429:
                wait = min(retry_after if retry_after is not None else 0, MAX_RETRY_AFTER)
                if wait > 0:
                    time.sleep(wait)
                logger.debug("crashlens: rate limited (429), dropping event")
                return
            if 500 <= status < 600:
                if attempt == 0:
                    time.sleep(RETRY_DELAY)
                    continue
                logger.debug("crashlens: server error %s, dropping event", status)
                return
            # 2xx success or a 4xx we cannot fix: done either way.
            return

    def _post(
        self, body: bytes, headers: Dict[str, str]
    ) -> "tuple[int, Optional[int]]":
        req = urllib.request.Request(
            self._url, data=body, headers=headers, method="POST"
        )
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                resp.read()
                return resp.status, None
        except urllib.error.HTTPError as exc:
            retry_after = _parse_retry_after(exc.headers.get("Retry-After"))
            try:
                exc.read()
            except Exception:  # noqa: BLE001 - draining is best-effort
                pass
            return exc.code, retry_after


def _parse_retry_after(value: Optional[str]) -> Optional[int]:
    if not value:
        return None
    try:
        return max(0, int(float(value.strip())))
    except (ValueError, AttributeError):
        return None

File: sdks/node/src/transport.ts
// Event transport. A bounded array queue feeds a single in-flight send loop.
// One event is POSTed per request over node:http / node:https.
//
// Design (matches the sibling SDKs and docs/PROTOCOL.md):
//   - The host app never blocks: send() enqueues and returns; the drain loop
//     runs on its own.
//   - Bounded queue: at maxQueue the OLDEST event is dropped (drop-oldest).
//   - gzip the body when it exceeds GZIP_THRESHOLD (4 KiB).
//   - 2s request timeout.
//   - On a socket error or 5xx, retry ONCE after RETRY_DELAY_MS, then drop.
//   - On 429, wait min(Retry-After, MAX_RETRY_AFTER_MS) then drop (no resend).
//   - Nothing here ever throws into the host application.

import * as http from "node:http";
import * as https from "node:https";
import { gzipSync } from "node:zlib";
import type { CrashlensEvent } from "./types";
import { delay } from "./util";

export const GZIP_THRESHOLD = 4096; // 4 KiB
export const RETRY_DELAY_MS = 500;
export const MAX_RETRY_AFTER_MS = 5000;
export const DEFAULT_TIMEOUT_MS = 2000;
export const DEFAULT_MAX_QUEUE = 100;

const SDK_USER_AGENT = "crashlens-node";

interface PostResult {
  status: number;
  retryAfterMs: number;
}

export class Transport {
  private readonly url: string;
  private readonly key: string;
  private readonly maxQueue: number;
  private readonly timeoutMs: number;
  private readonly lib: typeof http | typeof https;
  private queue: string[] = [];
  private sending = false;
  private closed = false;
  private warnedFull = false;

  constructor(
    url: string,
    key: string,
    maxQueue: number = DEFAULT_MAX_QUEUE,
    timeoutMs: number = DEFAULT_TIMEOUT_MS,
  ) {
    this.url = url;
    this.key = key;
    this.maxQueue = Math.max(1, maxQueue);
    this.timeoutMs = timeoutMs;
    this.lib = url.startsWith("https:") ? https : http;
  }

  send(event: CrashlensEvent): void {
    if (this.closed) return;
    let body: string;
    try {
      body = JSON.stringify(event);
    } catch {
      return;
    }
    this.enqueue(body);
    void this.drain();
  }

  // Drain the queue with a deadline. Returns true if the queue emptied.
  async flush(timeoutMs = 5000): Promise<boolean> {
    const deadline = Date.now() + timeoutMs;
    void this.drain();
    while (this.queue.length > 0 && !this.closed && Date.now() < deadline) {
      await delay(20);
    }
    return this.queue.length === 0;
  }

  close(): void {
    this.closed = true;
    this.queue = [];
  }

  // Exposed for tests.
  queueLength(): number {
    return this.queue.length;
  }

  private enqueue(body: string): void {
    this.queue.push(body);
    while (this.queue.length > this.maxQueue) {
      this.queue.shift(); // drop oldest
      if (!this.warnedFull) {
        this.warnedFull = true;
        // A single, quiet warning so the drop is discoverable without noise.
        if (typeof process !== "undefined" && process.emitWarning) {
          process.emitWarning(
            "crashlens: event queue is full; dropping oldest events",
          );
        }
      }
    }
  }

  private async drain(): Promise<void> {
    if (this.sending) return;
    this.sending = true;
    try {
      while (this.queue.length > 0 && !this.closed) {
        const body = this.queue[0];
        await this.transmit(body);
        this.queue.shift();
      }
    } catch {
      // Never let a transport failure escape into the host app.
    } finally {
      this.sending = false;
    }
  }

  private prepare(body: string): {
    data: Buffer;
    headers: Record<string, string | number>;
  } {
    let data = Buffer.from(body, "utf-8");
    const headers: Record<string, string | number> = {
      "Content-Type": "application/json; charset=utf-8",
      "X-Crashlens-Key": this.key,
      "User-Agent": SDK_USER_AGENT,
    };
    if (data.length > GZIP_THRESHOLD) {
      data = gzipSync(data);
      headers["Content-Encoding"] = "gzip";
    }
    headers["Content-Length"] = data.length;
    return { data, headers };
  }

  private async transmit(body: string): Promise<void> {
    const { data, headers } = this.prepare(body);
    for (const attempt of [0, 1]) {
      let result: PostResult;
      try {
        result = await this.post(data, headers);
      } catch {
        // Socket-level failure: retry once, then drop.
        if (attempt === 0) {
          await delay(RETRY_DELAY_MS);
          continue;
        }
        return;
      }
      if (result.status === 429) {
        const wait = Math.min(result.retryAfterMs, MAX_RETRY_AFTER_MS);
        if (wait > 0) await delay(wait);
        return; // drop, never resend on 429
      }
      if (result.status >= 500 && result.status < 600) {
        if (attempt === 0) {
          await delay(RETRY_DELAY_MS);
          continue;
        }
        return;
      }
      // 2xx success or a 4xx we cannot fix: done either way.
      return;
    }
  }

  private post(
    data: Buffer,
    headers: Record<string, string | number>,
  ): Promise<PostResult> {
    return new Promise<PostResult>((resolve, reject) => {
      const req = this.lib.request(
        this.url,
        { method: "POST", headers },
        (res) => {
          const retryAfterMs = parseRetryAfter(res.headers["retry-after"]);
          res.on("data", () => {
            // Drain the body so the socket can be reused / freed.
          });
          res.on("end", () => {
            resolve({ status: res.statusCode ?? 0, retryAfterMs });
          });
          res.on("error", reject);
        },
      );
      req.on("error", reject);
      req.setTimeout(this.timeoutMs, () => {
        req.destroy(new Error("crashlens: request timed out"));
      });
      req.write(data);
      req.end();
    });
  }
}

function parseRetryAfter(header: string | string[] | undefined): number {
  if (!header) return 0;
  const value = Array.isArray(header) ? header[0] : header;
  const secs = parseInt(value, 10);
  return Number.isFinite(secs) && secs > 0 ? secs * 1000 : 0;
}
